Replace Products (Beta) with Product Collection in product archive templates

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR replaces Products (Beta) block with Product Collection:
- in blockified product archive templates,
- in process of transforming Classic product archive templates with blocks.
In addition it adjusts the E2E tests and add more.

Additional minor improvements:
- Add Footer to blockified Products by Tag template
- Extract No Results content to pattern so it can be translated

Also, I adjusted `phpcs.xml` config in [here](https://github.com/woocommerce/woocommerce/pull/48112/files#diff-8b0d1360c2b971459712cf994fe069d2be570a950a412f2a73349ed9f454797d) excluding `woocommerce/patterns` directory from two rules that should not apply there. That was not passing with newly added file [`plugins/woocommerce/patterns/no-products-found-filters.php`](https://github.com/woocommerce/woocommerce/pull/48112/files#diff-8c0377c1723fcf4affacc44f46da4bf329a18054fc9e5e5bb20527ad749f1775) which has the same structure as any other pattern in the directory but linting job only compares with `origin/trunk` hence other files are passing linting even though they would fail otherwise. This change follows the config from `woocommerce-blocks` in [here](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/phpcs.xml#L47):

This PR will be merged with https://github.com/woocommerce/woocommerce/pull/48204 that hides Products (Beta) from inserter.

Closes https://github.com/woocommerce/woocommerce/issues/43411

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Default block

1. Go to Editor -> Product Catalog
2. Before entering the edit mode, click the three dots menu and "Clear customizations"

<img width="464" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/6e3fbb04-0d24-42d1-818d-c6e23106efa5">

3. Enter edit mode
4. Open tree view
5. Expected (check screenshot below):
    1. There's Product Collection instead of Products (Beta) block (1)
    2. Product Collection renders correctly as on screenshot (2)
    3. Sync query with template is enabled by default (3)

![Screen Shot 2024-06-05 at 15 50 33 PM](https://github.com/woocommerce/woocommerce/assets/20098064/26bf53b0-0a2a-4230-9885-7e18f273e332)

Repeat for templates:
- Products by Attribute
- Products by Category
- Products by Tag
- Product Search Results

For Product Search Results check additionally that "No Results" block includes the Product Search block. Due to another issue https://github.com/woocommerce/woocommerce/pull/48112, it may be that this will not load when the template is pristine. If that's the case please:
- make some edits to the Product Search Result Template (e.g. add paragraph with some text) and save changes
- clear customizations:

![image](https://github.com/woocommerce/woocommerce/assets/20098064/04214c88-e579-4783-a89a-3b7865730d1d)

In either of these two cases you should see this content in No Results block:

<img width="1365" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/454ccda3-04f9-4e82-8df5-5c64e2671139">


##### Before and after comparison of every template 
⚠️ Note: there should be no visual regression. Minor UI differences or copy differences are ok but feature-wise there should be no difference.

Template | Before - with Products (Beta) | After - with Product Collection
-- | -- | --
Product Catalog | <img width="1652" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/a367c9b9-b792-4d00-9fe0-afc73225c69d"> | <img width="1652" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/e7619aa3-51d3-48a8-96c3-dd32bde8fb30">
Products by Attribute | <img width="1652" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/a7a26bb6-d35c-41f9-9987-6431241f1759"> | <img width="1652" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/e5f965fc-1496-4b46-a253-d92b608c8425">
Products by Category | <img width="1655" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/553a4c65-e1e1-4297-b322-388b6a3832cf"> | <img width="1653" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/f02409a3-4ec2-4685-b5df-3e81b23f7eaa">
Products by Tag | <img width="1653" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/162ff22d-ca10-49b9-b049-03d7b47bc708"> | <img width="1653" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/9add5ad7-13de-4f9e-908c-e5aa2e434524">
Product Search Results | <img width="1654" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/6d0a818e-6ad8-46e8-ad6c-e20c3204eaab"> | <img width="1493" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/06728d4e-87e0-4605-8bbb-7167815b00b4">


#### Transforming classic templates

1. Go to Editor -> Product Catalog
2. Remove the content
3. Add "Product Grid (Classic)" block
4. Hover over it and click "Transform into blocks"
5. **Expected:** the template is transformed into blocks and:
    1. There's Product Collection in the blocks structure
    2. Product Collection renders correctly
    3. Sync query with template is enabled

Repeat for templates:
- Products by Attribute (use "Product Attribute (Classic)" block)
- Products by Category (use "Product Category (Classic)" block)
- Products by Tag (use "Product Tag (Classic)" block)
- Product Search Results (use "Product Search Results (Classic)" block)

For Product Search Results check additionally that "No Results" block includes the Product Search block:

<img width="1231" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/66fbbfbb-e2e1-4a58-a8a9-d6362bb90762">

##### Before and after comparison of every template after clicking "Transform into blocks"
⚠️ Note: there should be no visual regression. Minor UI differences or copy differences are ok but feature-wise there should be no difference.

Template | Before - with Products (Beta) | After - with Product Collection
-- | -- | --
Product Catalog | <img width="1654" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/3d388a56-50ed-423a-bbfb-6cd02ac9d8b3"> | <img width="1656" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/f2a8ab61-6bc7-4537-861b-18f15a73d409">
Products by Attribute | <img width="1656" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/cd657a99-d1fd-4682-9715-62a3341c7d63"> | <img width="1655" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/fbc05de6-17a6-451b-9ce0-384535d9d317">
Products by Category | <img width="1654" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/a539a25c-5b51-4e57-900d-ab158f23c1ed"> | <img width="1653" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/fbeb44d3-8ee5-412b-a596-d6c5c3ff1d15">
Products by Tag | <img width="1655" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/e9abedef-a052-4efd-8979-3794279f83eb"> | <img width="1655" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/ad9944cb-8d00-4073-9577-b9c6866606fb">
Product Search Results | <img width="1653" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/14fa08b5-9d75-42e1-aaff-3b2539ecbfe6"> | <img width="1656" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/36143a80-60a3-461b-a3cc-702eb6b8483f">

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->


</details>

<details>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->


</details>

kmanijak committed:
Replace Products (Beta) with Product Collection block in blockified template

kmanijak committed:
Add necessary attributes that influences the look of blocks

kmanijak committed:
Replace Products (Beta) with Product Collection block in product search results

kmanijak committed:
Replace Products (Beta) with Product Collection block in products by attribute

kmanijak committed:
Replace Products (Beta) with Product Collection block in products by category

kmanijak committed:
Replace Products (Beta) with Product Collection block in products by tag

kmanijak committed:
Replace Products (Beta) with Product Collection when transforming from classic product archive

kmanijak committed:
Change the no results content in blockified search results template

kmanijak committed:
Replace Products (Beta) with Product Collection block in product search results

kmanijak committed:
Add product search in no results content of blockified product search result

kmanijak committed:
Add ts-ignore before createBlocksFromInnerBlocksTemplate imports

kmanijak committed:
Add changelog

kmanijak committed:
Remove step of replacing Products (Beta) with Product Collection in compatibility layer tests

kmanijak committed:
Remove step of replacing Products (Beta) with Product Collection in Product Collection tests

kmanijak committed:
Fix lint

kmanijak committed:
Update E2E tests of Product Collection

kmanijak committed:
Adjust Products (Beta) E2E tests

kmanijak committed:
Add Product Collection focus step in E2E tests

kmanijak committed:
Improve the way PC block is focused

kmanijak committed:
Further tests adjustments and new tests comparing classic template with Product Collection

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/products/products.block_theme.spec.ts:
```diff
@@ -65,14 +66,12 @@ test.describe( `${ blockData.name } Block `, () => {
-		await editor.canvas.locator( 'body' ).click();
-
+		await editor.enterEditMode();
+		await editor.setContent( '' );
```
tip: Creating a new custom `archive-product` template would be more efficient as it wouldn't need to wait for the whole template to load before clearing it. Like in this test:

https://github.com/woocommerce/woocommerce/blob/460d73eee07e4ca28954f33d509d7c1414ad1f5b/plugins/woocommerce-blocks/tests/e2e/tests/add-to-cart-form/add-to-cart-form.block_theme.spec.ts#L85-L97

kmanijak replied:
👍 Addressed!

kmanijak replied:
I had to bring back the old logic as this specific test logic requires an actual product catalog template.

WunderBart replied:
I guess the `slug: 'product-catalog'` part should actually be `slug: 'archive-product'` so that it overwrites the original template and uses the correct logic. 🤔

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.page.ts:
```diff
@@ -606,6 +601,11 @@ class ProductCollectionPage {
 
+	async getProductNames() {
+		const products = this.page.locator( '.wp-block-post-title' );
+		return products.allTextContents();
```
We need to be careful when getting text content this way as it's inherently flaky. To address this, we should assert the number of elements that we expect to be rendered before getting their content. If we don't do it, we will only get text content from the elements that were rendered at the time this method was called.

kmanijak replied:
Good tip! In this case, the products are server-side-rendered so I expect all of them to be visible from the very beginning but I added assertion anyway 👍

kmanijak committed:
Bring back the click on body

kmanijak committed:
Include tips from reviewer

kmanijak committed:
Enter edit mode, step that I accidentally removde

kmanijak committed:
Bring the E2E test flow with updating product catalog

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts:
```diff
@@ -803,6 +803,7 @@ test.describe( 'Product Collection', () => {
 				canvas: 'edit',
 			} );
 
+			await editor.enterEditMode();
```
If the the `visitSiteEditor` has the `canvas: 'edit'` param set, it will load the entity in edit mode, which might cause a conflict with `enterEditMode()`. I'd generally recommend using `enterEditMode()` for now, since it also waits until the editor is loaded (meaning that I would just remove the `canvas: 'edit'` part 😄).

kmanijak committed:
Change the way of removing focus from PC block in tests

kmanijak committed:
Fix lint

kmanijak committed:
Improve the layout of search in No Results

kmanijak committed:
Use previous patterns in Product Search No Results to preserve translations

kmanijak committed:
Add Search button label in product search form pattern

kmanijak committed:
Merge branch 'trunk' into 43411-product-collection-replace-products-beta-with-product-collection-in-product-archive-templates

kmanijak committed:
Fix lint

kmanijak committed:
Improve comment explanation

kmanijak committed:
Extract No Results content to separate pattern in order to make the content translatable

kmanijak committed:
Lint fix

kmanijak committed:
Add missing footer to the blockified Products by Tag template

kmanijak committed:
Exclude woocommerce/patterns from phpcs rules that dont apply there

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/cart-checkout-shared/use-forced-layout/index.ts:
```diff
@@ -3,13 +3,14 @@
 import {
 	createBlock,
 	getBlockType,
+	// @ts-expect-error Type definitions for this function are missing in Guteberg
```
```suggestion
	// @ts-expect-error Type definitions for this function are missing in Gutenberg
```

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/classic-template/archive-product.ts:
```diff
@@ -1,38 +1,43 @@
+} from '@woocommerce/blocks/product-collection/constants';
 import {
 	createBlock,
+	// @ts-expect-error Type definitions for this function are missing in Guteberg
```
```suggestion
	// @ts-expect-error Type definitions for this function are missing in Gutenberg
```

imanish003 reviewed (commented):
Hey Karol! Thank you for the detailed PR description 🚀 
I haven't reviewed the code yet. Here is what found during my functional testing. I am going to review the code now. 🤝

> Repeat for templates:
> 
> - Products by Attribute
> - Products by Category
> - Products by Tag
> - Product Search Results

- I noticed that "Responsive" toggle isn't ON by default in the Product Collection block. I think it should be ON by default. What do you think?

> ⚠️ Note: there should be no visual regression. Minor UI differences are ok but feature-wise or content-wise there should be no difference

I don't think it's a problem but just wanted to highlight that "No Results" block is different between Products (Beta) and Product Collection. 

> Products by Tag	

In "Products (Beta)" block, "No results" block is empty, but in Product Collection it isn't. I think it's better to have content in the No Results block. What do you think? I hope it won't have any issues because we have to maintain backward compatibility.

> For Product Search Results check additionally that "No Results" block includes the Product Search block:

In Products beta:
```
<!-- wp:search {"showLabel":false,"placeholder":"Search products…","query":{"post_type":"product"}} /-->
```

In Product Collection:
```
<!-- wp:search {"showLabel":false,"placeholder":"Search products…","buttonText":"Search","query":{"post_type":"product"}} /-->
```

There is one difference. It was missing buttonText in the Products beta block. Maybe this was an issue in the older template. I think it makes sense to have button text. What do you think?

> Before and after comparison of every template after clicking "Transform into blocks"

- In this case, the Responsive toggle is ON by default in the Product Collection block. This isn't the case with template content as I raised earlier. I think it should be consistent. What do you think?

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/cart-checkout-shared/use-forced-layout/index.ts:
To confirm, changes in this file are to fix TS error only. Right?

kmanijak replied:
Exactly!

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/classic-template/product-search-results.ts:
```diff
@@ -43,18 +44,23 @@ const extendInnerBlocksWithNoResultsContent = (
 	innerBlocks: InnerBlockTemplate[],
 	inheritedAttributes: InheritedAttributes
 ) => {
+	// InnerBlockTemplate is an array block representation so properties
+	// like name or attributes need to be accessed with array indexes.
+	const nameArrayIndex = 0;
+	const attributesArrayIndex = 1;
+
```
Thanks for creating these constants. It makes the code more readable. 🙌🏻

imanish003 commented on plugins/woocommerce/patterns/product-search-form.php:
```diff
@@ -6,4 +6,5 @@
 ?>
-<!-- wp:search {"showLabel":false,"placeholder":"<?php echo esc_attr_x( 'Search products…', 'placeholder for search field', 'woocommerce' ); ?>","query":{"post_type":"product"}} /-->
+
+<!-- wp:search {"showLabel":false,"placeholder":"<?php echo esc_attr_x( 'Search products…', 'placeholder for search field', 'woocommerce' ); ?>","buttonText":"<?php echo esc_attr_x( 'Search', 'button label of product search block', 'woocommerce' ); ?>","query":{"post_type":"product"}} /-->
```
Thanks for adding using `esc_attr_x`. This is a good small improvement 🙌

imanish003 reviewed (commented):
I reviewed the code and it looks good to me 🚀 
I will wait for your reply to my previous comments before approving the PR.

kmanijak committed:
Make PC block responsive by default in product archive templates

kmanijak committed:
Fix typo

kmanijak commented:
Thanks @imanish003 for reviewing and testing!

> I noticed that "Responsive" toggle isn't ON by default in the Product Collection block. I think it should be ON by default. What do you think?

Good catch, enabled! :)

> I don't think it's a problem but just wanted to highlight that "No Results" block is different between Products (Beta) and Product Collection.

Yes, it's been updated to what we use in Product Collection. I updated the notice in PR description.

> In "Products (Beta)" block, "No results" block is empty, but in Product Collection it isn't. I think it's better to have content in the No Results block. What do you think? I hope it won't have any issues because we have to maintain backward compatibility.

I'm sure it was a bug that No Results had no content in this template. But what issues do you have in mind regarding backward compatibility? 

> There is one difference. It was missing buttonText in the Products beta block. Maybe this was an issue in the older template. I think it makes sense to have button text. What do you think?

Yeah, I purposefully added it to `plugins/woocommerce/patterns/product-search-form.php` as I'm pretty sure it was missing by accident. Do I understand correctly you agree with adding it?

---

I think I addressed all of the comments, may I ask you for re-reviewing? 🙏

imanish003 committed:
Fix typo

imanish003 reviewed (approved):
Hey Karol! Thanks for addressing the feedback. I am approving this PR as it looks good to me. Well Done 🚀

> Good catch, enabled! :)

Awesome. I verified that the "Responsive" toggle is ON by default 🚀


> I'm sure it was a bug that No Results had no content in this template. But what issues do you have in mind regarding backward compatibility?

For example, there was nothing in the No Results block in the Products (Beta) block. Now when user will update WooCommerce version, they will see content in the No Results block (Editor as well Frontend). I believe this is a good change. I just wanted to confirm if you see any issues with this change.

> Yeah, I purposefully added it to plugins/woocommerce/patterns/product-search-form.php as I'm pretty sure it was missing by accident. Do I understand correctly you agree with adding it?

Yes, I agree with adding it. It's a good change.
